sqlmesh/migrations/v0078_detect_diff_caused_py_metadata_flag_propagation_and_warn.py

```diff
@@ -0,0 +1,491 @@
+
+                for property_name in PROPERTIES:
+                    if property_value := node.get(property_name):
+                        properties = parse_properties(property_value, dialect)
+                        if isinstance(properties, exp.Tuple):
+                            statements.extend(properties.expressions)
+
+                jinja_macro_references, _ = extract_macro_references_and_variables(
+                    *(gen(e) for e in statements)
+                )
+
+                if audit_definitions := node.get("audit_definitions"):
+                    audit_queries = [
+                        parse_expression(audit["query"], audit["dialect"])
+                        for audit in audit_definitions.values()
+                    ]
+                    statements.extend(audit_queries)
+
+                for _, audit_args in func_call_validator(node.get("audits") or []):
+                    statements.extend(audit_args.values())
+
+                metadata_func_candidates = set()
+
+                for signal_name, signal_args in func_call_validator(
+                    node.get("signals") or [], is_signal=True
+                ):
+                    metadata_func_candidates.add(signal_name)
+                    statements.extend(signal_args.values())
+
+                metadata_func_candidates |= extract_used_macros(statements, jinja_macro_references)
+                if metadata_func_candidates and any(
+                    c in python_env for c in metadata_func_candidates
+                ):
+                    serialized_python_env = {k: Executable(**v) for k, v in python_env.items()}
+                    hydrated_python_env = prepare_env(serialized_python_env)
+
+                    for func_name in metadata_func_candidates:
+                        func = hydrated_python_env.get(func_name)
+                        if not func:
+                            continue
+
+                        executable = serialized_python_env[func_name]
+                        if executable.is_metadata and any(
+                            global_ref in hydrated_python_env
+                            for global_ref in func_globals(func, executable.payload)
+                        ):
+                            get_console().log_warning(
+                                f"Node '{node['name']}' references a metadata-only function (macro or signal), "
+                                "which in turn relies on non-metadata objects. This means that the next "
+                                "plan command may detect unexpected changes and prompt about backfilling "
+                                "this model. If this is a concern, consider running a forward-only plan "
+                                "instead: https://sqlmesh.readthedocs.io/en/stable/concepts/plans/#forward-only-plans.\n"
+                            )
+                            break
+        except Exception:
+            pass
+
+        new_snapshots.append(
```

Comment:
do you mutate snapshots at all? why are you reinserting them?

Reply:
good point, I had started out with the goal of mutating them but I forgot to change this 👍